lexer/lexer.hpp

```diff
@@ -1,3 +1,121 @@
 #pragma once
 
-int sum(int a, int b);
+#include "trie.hpp"
+#include "token.hpp"
+#include <cctype>
+
+namespace lexer {
+
+extern common::Trie<TokenType> g_keywordTrie;
+extern common::Trie<TokenType> g_operatorTrie;
+
+class Lexer {
+public:
+    Lexer(std::string src) : m_buf(src) {}
+
+    Token Next() {
+
+        // Ignore spaces
+        for (; m_pos < m_buf.size() && Lexer::isspace(m_buf[m_pos]); m_pos++) {}
+
+        if (m_pos == m_buf.size()) {
+            return { .type = TokenType::Eof };
+        }
+
+        Token tok{
+            .type = TokenType::Error,
+            .line = m_lineNum,
+            .srcPos = m_pos - m_lastAfterNewLine,
+        };
+
+        // If alpha -> either a keyword or identifier -> read until space
+        if (Lexer::isidstart(m_buf[m_pos])) {
+            size_t len = 1;
+            for (; m_pos + len < m_buf.size() && Lexer::isidsuf(m_buf[m_pos + len]); len++) {}
+
+            tok.image = m_buf.substr(m_pos, len);
+            tok.type = g_keywordTrie.Find(tok.image).value_or(TokenType::Identifier);
+ 
+            m_pos += len;
+
+            return tok;
+        }
+
+        if (std::isdigit(m_buf[m_pos])) {
+            size_t truncLen = 1;
+            for(; m_pos + truncLen < m_buf.size() && std::isdigit(m_buf[m_pos + truncLen]); truncLen++) {}
+
+            size_t fracLen = 0;
+            if (m_pos + truncLen < m_buf.size() && m_buf[m_pos + truncLen] == '.') {
```

Comment:
Should there be something to ensure that the number is followed by a special character and not a letter or underscore?

Reply:
If you're talking about constructs like `long long foo = 4ul`, then no.

Reply:
Why not? an identifier/keyword can stick to a number from the left? This would also prevent us from allowing constructs like the one you wrote in the future.

Reply:
Okay, I see what you mean. But on the other hand, stopping tokenizing a number when a non-number symbol is encountered is the simplest solution for this. Hence, I think, there should be no problem at changing this behavior if we need to in the future.